File: automation/features/support/step_definitions/gallery_steps.js
var gallery = require('../page_objects/gallery_widget');
var world = require('../world');
var wait = require('../utils/wait');
var findValue = require('../utils/findValue');

module.exports = function() {

    this.Given(/^I can see the logo on the gallery header$/, function() {
        browser.waitForVisible(gallery.headerLogo, 10000);
        expect(browser.isVisible(gallery.headerLogo)).toBe(true);
    });

    this.Given(/^I can click the logo to go to homepage$/, function() {
        var logoLink = browser.getAttribute(gallery.headerLogo, 'href');
        expect(logoLink).toEqual(world.Urls.home_page);
    });

    this.Given(/^I can see an image appearing on the gallery$/, function() {
        var img = browser.getAttribute(gallery.galleryImg, 'src');
        expect(img).not.toBe(null);
        console.log("IMAGE SRC =" + " " + img);
    });

    this.Given(/^I can see the source appearing on the gallery with gtm "([^"]*)"$/, function (gtm) {
        //Get values
        var sourceHref = browser.getAttribute(gallery.gallerySource, 'href');
        var sourceGTM = browser.getAttribute(gallery.gallerySource,'class');
        var sourceLogo = browser.getAttribute(gallery.gallerySourceImg,'src');

        //Validate the values
        console.log(sourceHref);
        expect(sourceHref).not.toEqual('');
        console.log(sourceGTM);
        expect(sourceGTM).toEqual(gtm);
        console.log(sourceLogo);
        expect(sourceLogo).not.toEqual('');
    });

    this.Given(/^I can see the created date on the gallery "([^"]*)"$/, function(date) {
        wait(2000);
        browser.waitForExist(gallery.galleryDate,5000);
        browser.scroll(gallery.galleryDate);
        var galleryDate = browser.getText(gallery.galleryDate);
        console.log(galleryDate);
        expect(galleryDate).toContain(date);
        console.log('gallery date is:' + galleryDate)
    });

    this.Given(/^I can see the gallery title containing "([^"]*)"$/, function(longTitle) {
        var galleryTitle = browser.getText(gallery.galleryLongTitle);
        expect(galleryTitle).toContain(longTitle);
    });

    this.Given(/^I can not see the gallery title$/, function() {
        expect(browser.isVisible(gallery.galleryLongTitle)).toBe(false);
    });

    this.Given(/^I should see the long title on the gallery header on the next gallery slide$/, function() {
        expect(browser.isVisible(gallery.galleryLongTitle)).toBe(true);
    });

    this.Given(/^I should not see the long title on the gallery header on the next gallery slide$/, function() {
        expect(browser.isVisible(gallery.galleryLongTitle)).toBe(false);
    });

    this.Given(/^I can see the gallery description of the gallery containing "([^"]*)"$/, function(description) {
        var galleryDescription = browser.getText(gallery.galleryDescription);
        expect(galleryDescription).toContain(description);
    });

    this.Given(/^I can see the image number "([^"]*)" of total "([^"]*)" on the gallery$/, function(num, total) {
        var imageCount = browser.getText(gallery.imageCount);
        expect(imageCount[0]).toEqual(num + " / " + total);
        console.log(imageCount[0]);
    });

    this.Given(/^I can see the image caption on the gallery containing "([^"]*)"$/, function(caption) {
        browser.waitForVisible(gallery.imgCaption);
        var imgCaption = browser.getText(gallery.imgCaption);
        expect(imgCaption[0]).toMatch(caption);
        console.log(imgCaption[0]);
    });

    this.When(/^I see the image no "([^"]*)" on the gallery$/, function(imgNum) {
        expect(browser.getText(gallery.currentImgNum)).toMatch(imgNum);
    });

    this.When(/^I see the video ID "([^"]*)" on the gallery$/, function(videoId) {
        browser.waitForVisible(gallery.videoWrapper, 3000);
        browser.moveToObject(gallery.videoWrapper);
        expect(browser.getAttribute(gallery.videoWrapper, gallery.videoId)).toEqual(videoId)
    });

    this.When(/^I can see the play button and click on it$/, function() {
        browser.waitForVisible(gallery.playButton, 3000);
        browser.click(gallery.playButton);
        expect(browser.isVisible(gallery.videoPlayWrap, gallery.videoAdPlay)).toBe(true);
    });

    this.When(/^I can see the facebook share button on gallery page$/, function () {
        var facebook = browser.isVisible(gallery.galleryFacebook);
        var facebookButton = browser.getText(gallery.galleryFacebook);
        expect(facebook).toBe(true);
        expect(facebookButton).toEqual('SHARE');
    });

    this.When(/^I can see the pinterest share button on gallery page$/, function () {
        var pinterest = browser.isVisible(gallery.galleryPinterest);
        var pinterestButton = browser.getText(gallery.galleryPinterest);
        expect(pinterest).toBe(true);
        expect(pinterestButton).toEqual('PIN');
    });

    this.Given(/^I can see the author "([^"]*)" on the gallery$/, function (authorName) {
        var author = browser.getText(gallery.authorText);
        expect(author).toContain(authorName);
    });
};

File: automation/dolly/features/step_definitions/gallery_steps.rb
Given(/^I can see the dolly logo on the gallery header$/) do
    expect(find('.icon-logo').native.style('background-image')).to match("/assets/images/dolly-logo.svg")
    # Validate the dolly logo on the gallery header
end

Given(/^I can click the dolly logo to go to homepage$/) do
    expect(find('.icon-logo')[:href].to_s).to eq($base_url)
    # Validate the dolly logo is clickable to open Homepage
end

Given(/^I can see the long title on the gallery header "([^"]*)"$/) do |long_title|
    gallery_long_title = find('.gallery__title').text
    expect(gallery_long_title).to match(long_title)
    puts (gallery_long_title)
    # Validate the long title on the gallery header
end

Given(/^I should not see the long title on mobile$/) do
    expect(page).to_not have_selector('.gallery__title')
    # Validate the long title should not appear on mobile
end

Given(/^I can see an image appearing on the gallery$/) do
    within find('.gallery__slide') do
        gallery_image = find('img')[:src].to_s
        expect(gallery_image).to_not eq("")
        puts "<img src='" + gallery_image + "' style='width:5%;height:5%;'></img>"
    end
    # Validate the first image of the gallery
end

Given(/^I can see the custom label of the gallery$/) do
    within find('.gallery__summary') do
        # waiting DDO-305
    end
    pending # completing this step once DDO-305 is done
end

Given(/^I can see the created date of the gallery "([^"]*)"$/) do |date|
    gallery_date = find('.gallery__date').text
    expect(gallery_date).to eq(date) #This need to be updated once DDO-306 is fixed
    puts (gallery_date)
    # Validate the created date
end

Given(/^I can see the created date of this gallery$/) do
    gallery_date = find('.gallery__date').text
    expect(gallery_date).to_not eq("")
    puts (gallery_date)
    # Validate the created date
end

Given(/^I can see the gallery description of the gallery containing "([^"]*)"$/) do |message|
    gallery_description = find('.gallery__summary-text').text
    expect(gallery_description).to match(message)
    # Validate the gallery description
end

Given(/^I should not see the gallery description on mobile$/) do
    expect(page).to_not have_selector('.gallery__summary-text')
    # Validate the gallery description should not appear on mobile
end

Given(/^I should not see the left arrow on the gallery$/) do
    expect(page).to_not have_selector('button.gallery__nav.gallery__nav--prev')
    # Validate the left arrow should not appear
end

When(/^I can see the right arrow on the gallery$/) do
    find('button.gallery__nav.gallery__nav--next')
    # Validate the right arrow should appear on the gallery
end

Given(/^I can see the image number "([^"]*)" of total "([^"]*)" on the gallery$/) do |no,total|
    gallery_image_count = find('.gallery__slide-count').text
    expect(gallery_image_count).to eq(no + " / " + total)
    puts (gallery_image_count)
    # Validate the count image number
end

Given(/^I can see the image caption on the gallery containing "([^"]*)"$/) do |message|
    gallery_image_caption = find('.gallery-caption__content').text
    expect(gallery_image_caption).to match(message)
    puts (gallery_image_caption)
    # Validate the image caption on the gallery
end

Given(/^I can click MORE to see the full image caption on the gallery$/) do
    find('.gallery-caption__toggle--more').click
    expect(find('.gallery-caption__toggle--less')[:style].to_s).to eq("opacity: 1;")
    # Validate the more button in the image caption
end

Given(/^I can click LESS to see the short image caption on the gallery$/) do
    expect(page).to_not have_selector('.gallery-caption__toggle--more')
    find('.gallery-caption__toggle--less').click
    find('.gallery-caption__toggle--more')
    # Validate the less button in the image caption
end

Given(/^I can click the right arrow on the gallery$/) do
    current_image_number = find('.gallery__slide-current').text.to_i
    find('button.gallery__nav.gallery__nav--next').click
    next_image_number = find('.gallery__slide-current').text.to_i
    expect(next_image_number).to be > (current_image_number)
    # Validate the right arrow on the gallery can show next slide
end

Given(/^I can click the right arrow on the gallery "([^"]*)" times$/) do |num|
    for i in 1..num.to_i
        find('button.gallery__nav.gallery__nav--next').click
    end
end

Given(/^I can click the right arrow on the gallery on the last slide$/) do
    find('button.gallery__nav.gallery__nav--next').click
    # Validate the right arrow on the gallery on the last slide
end

When(/^I see the image no "([^"]*)" on the gallery$/) do |no|
    expect(find('.gallery__slide-current').text).to eq(no)
    # Validate the page shows the correct image no
end

When(/^I can see the left arrow on the gallery$/) do
    find('button.gallery__nav.gallery__nav--prev')
    # Validate the left arrow appear on the gallery page
end

When(/^I can click the left arrow to go back to a previous image on the gallery$/) do
    current_image_number = find('.gallery__slide-current').text.to_i
    find('button.gallery__nav.gallery__nav--prev').click
    previous_image_number = find('.gallery__slide-current').text.to_i
    expect(previous_image_number).to be < (current_image_number)
    # Validate the left arrow on the gallery can show previous slide
end

When(/^I see the video ID "([^"]*)" on the gallery$/) do |id|
    # The video is in the third slide in stubbed data
    find('button.gallery__nav.gallery__nav--next').click
    find('button.gallery__nav.gallery__nav--next').click
    expect(find('.video-wrapper video')[:'data-video-id'].to_s).to eq(id)
    # Validate the video appearing on the slide
end

When(/^I can see the play button and click on it$/) do
    find('button',:text=>"Play Video").click
    find('.video-wrapper').hover
    find('.vjs-control-bar')
    # Validate the play button appearing and clickable
end

When(/^I can click the left and right arrows on the gallery$/) do
    find('button.gallery__nav.gallery__nav--prev').click
    find('button.gallery__nav.gallery__nav--next').click
    find('button.gallery__nav.gallery__nav--next').click
    find('button.gallery__nav.gallery__nav--prev').click
    # Validate that the right and left arrows are clickable on that slide
end

When(/^I see the last image on the gallery$/) do
    current_slide = find('.gallery__slide-current').text.to_i
    last_slide = find('.gallery__slide-last').text.to_i
    while current_slide != last_slide do
         case current_slide
             when 3,7,11,15 then  #to ensure we skip the ad slide
                 find('button.gallery__nav.gallery__nav--next').click
         end
         find('button.gallery__nav.gallery__nav--next').click
         current_slide = find('.gallery__slide-current').text.to_i
    end
    puts find('.gallery__slide-count').text
    # Validate the last slide on the gallery
end

When(/^I see the next gallery slide on the gallery$/) do
    expect(find('.gallery__next h5').text).to eq("NEXT GALLERY")
    # Validate the page is showing the next gallery slide
end

When(/^I see the next gallery slide on the gallery on mobile$/) do
    expect(find('.gallery__more-heading').native.style("background")).to match("/assets/images/more-galleries.svg")
    # Validate the page is showing the next gallery slide
end

When(/^I can see the next gallery name "([^"]*)"$/) do |name|
    expect(find('.gallery__next h2').text).to match(name)
    # Validate the next gallery name
end

When(/^I can see the next gallery name on mobile "([^"]*)"$/) do |nextGallery|
    expect(find('.gallery__next-main .teaser__title').text).to match(nextGallery)
    # Validate the next gallery name on mobile
end

When(/^I can see the right arrow on the next gallery slide$/) do
    find('.icon-arrow-next')
    # Validate the right arrow on the next gallery slide
end

Given(/^I can see an ad displayed$/) do
    find(".gallery__slide-ad ")
    #move to the next slide
    find('button.gallery__nav.gallery__nav--next').click
end
